src/read/util.rs

```diff
@@ -0,0 +1,158 @@
+pub(crate) unsafe trait Array {
+    type Item;
+    type Storage;
+
+    fn new_storage() -> Self::Storage;
+    fn as_slice(storage: &Self::Storage) -> &[MaybeUninit<Self::Item>];
+    fn as_mut_slice(storage: &mut Self::Storage) -> &mut [MaybeUninit<Self::Item>];
+}
+
+macro_rules! impl_array {
+    () => {};
+    ($n:literal $($rest:tt)*) => {
+        // SAFETY: does not modify the content in storage.
+        unsafe impl<T> Array for [T; $n] {
+            type Item = T;
+            type Storage = [MaybeUninit<T>; $n];
+
+            fn new_storage() -> Self::Storage {
+                // SAFETY: An uninitialized `[MaybeUninit<_>; _]` is valid.
+                unsafe { MaybeUninit::uninit().assume_init() }
+            }
+
+            fn as_slice(storage: &Self::Storage) -> &[MaybeUninit<T>] {
+                storage
+            }
+
+            fn as_mut_slice(storage: &mut Self::Storage) -> &mut [MaybeUninit<T>] {
+                storage
+            }
+        }
+        impl_array!($($rest)*);
+    }
+}
+
+impl_array!(4 192);
+
+pub(crate) struct ArrayVec<A: Array> {
+    storage: A::Storage,
+    len: usize,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub(crate) struct CapacityFull;
+
+impl<A: Array> ArrayVec<A> {
+    pub fn new() -> Self {
+        Self {
+            storage: A::new_storage(),
+            len: 0,
+        }
+    }
+
+    pub fn clear(&mut self) {
+        let ptr: *mut [A::Item] = &mut **self;
+        // Set length first so the type invariant is upheld even if `drop_in_place` panicks.
+        self.len = 0;
+        // SAFETY: `ptr` contains valid elements only and we "forget" them by setting the length.
+        unsafe { ptr::drop_in_place(ptr) };
+    }
+
+    pub fn try_push(&mut self, value: A::Item) -> Result<(), CapacityFull> {
+        let storage = A::as_mut_slice(&mut self.storage);
+        if self.len >= storage.len() {
+            return Err(CapacityFull);
+        }
+
+        storage[self.len] = MaybeUninit::new(value);
+        self.len += 1;
+        Ok(())
+    }
+
+    pub fn pop(&mut self) -> Option<A::Item> {
+        if self.len == 0 {
+            None
+        } else {
+            self.len -= 1;
+            // SAFETY: this element is valid and we "forget" it by setting the length.
+            Some(unsafe { A::as_slice(&mut self.storage)[self.len].as_ptr().read() })
+        }
+    }
+
+    pub fn swap_remove(&mut self, index: usize) -> A::Item {
+        assert!(self.len > 0);
+        A::as_mut_slice(&mut self.storage).swap(index, self.len - 1);
+        self.pop().unwrap()
+    }
+}
+
+impl<A: Array> Drop for ArrayVec<A> {
+    fn drop(&mut self) {
+        self.clear();
+    }
+}
+
+impl<A: Array> Default for ArrayVec<A> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl<A: Array> ops::Deref for ArrayVec<A> {
+    type Target = [A::Item];
+
+    fn deref(&self) -> &[A::Item] {
+        let slice = &A::as_slice(&self.storage)[..self.len];
```

Comment:
We didn't previously bounds check this in `RegisterRuleMap::rules` or `RegisterRuleMap::rules_mut`. There is a performance regression for the benchmark `cfi::iterate_entries_evaluate_every_fde`, and removing this and the bounds check in `deref_mut` seems to fix that.

Reply:
What's the status of this PR? Do you think this bounds check is something that can be changed?

Reply:
Sorry I missed the comment. Will fix and push soon.